lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	if !txMeta.RemoveYieldCurvePoint && txMeta.LockupYieldDurationNanoSecs > 0 {
+		if operationData.PrevLockupYieldCurvePoint == nil {
+			bav._deleteLockupYieldCurvePoint(&LockupYieldCurvePoint{
+				ProfilePKID:            profilePKID,
+				LockupDurationNanoSecs: txMeta.LockupYieldDurationNanoSecs,
+			})
+		} else {
+			bav._setLockupYieldCurvePoint(&LockupYieldCurvePoint{
+				ProfilePKID:               profilePKID,
+				LockupDurationNanoSecs:    operationData.PrevLockupYieldCurvePoint.LockupDurationNanoSecs,
+				LockupYieldAPYBasisPoints: operationData.PrevLockupYieldCurvePoint.LockupYieldAPYBasisPoints,
+			})
+		}
+	}
+
+	// Check if the transaction deleted a yield curve point. If it did, we add back the previous point.
+	// If the previous point is nil, we throw an error. This shouldn't be possible.
+	if txMeta.RemoveYieldCurvePoint && txMeta.LockupYieldDurationNanoSecs > 0 {
+		if operationData.PrevLockupYieldCurvePoint == nil {
+			return fmt.Errorf("_connectUpdateCoinLockupParams: trying to revert point deletion " +
+				"but found nil previous yield curve point; this shouldn't be possible")
+		}
+		bav._setLockupYieldCurvePoint(&LockupYieldCurvePoint{
+			ProfilePKID:               profilePKID,
+			LockupDurationNanoSecs:    operationData.PrevLockupYieldCurvePoint.LockupDurationNanoSecs,
+			LockupYieldAPYBasisPoints: operationData.PrevLockupYieldCurvePoint.LockupYieldAPYBasisPoints,
+		})
+	}
+
+	// Check if the transaction updated transfer restrictions. If it did, we reset the previous transfer restrictions.
+	if txMeta.NewLockupTransferRestrictions {
+		// Fetch the profile entry and LockupTransferRestriction status.
+		profileEntry := bav.GetProfileEntryForPKID(profilePKID)
+		if profileEntry == nil || profileEntry.isDeleted {
+			return fmt.Errorf("_connectUpdateCoinLockupParams: Trying to revert lockup transfer restriction " +
+				"update but found nil profile entry; this shouldn't be possible")
+		}
+
+		// Update the transfer restrictions.
+		profileEntry.DAOCoinEntry.LockupTransferRestrictionStatus = operationData.PrevLockupTransferRestriction
+		bav._setProfileEntryMappings(profileEntry)
+	}
+
+	// Decrement the operationIndex. We expect to find the basic transfer UtxoOps next.
+	operationIndex--
+	if operationIndex < 0 {
+		return fmt.Errorf("_disconnectUpdateCoinLockupParams: Trying to revert OperationTypeUpdateCoinLockupParams " +
+			"but found malformed utxoOpsForTxn")
+	}
+
+	// By here we only need to disconnect the basic transfer associated with the transaction.
+	basicTransferOps := utxoOpsForTxn[:operationIndex]
+	err := bav._disconnectBasicTransfer(currentTxn, txnHash, basicTransferOps, blockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectUpdateCoinLockupParams")
+	}
+	return nil
+}
+
+//
+// CoinLockupTransfer Transaction Logic
+//
+
+func (bav *UtxoView) _connectCoinLockupTransfer(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	verifySignatures bool,
+) (_totalInput uint64,
```

Comment:
Meta comment after reviewing this function. Initially it was such a strange thing to me. On the one hand, there's no real reason why this function shouldn't exist. It's kindof like baking liquid staking into the protocol level. But on the other hand, the whole point of lockups is that you CANT transfer them!

After noodling on it, I like this function. It effectively implements the "liquid bonding" concept that we talked about in Revolution, whereby locked coins can be sent back and forth similar to a "bond." I also like that it's optional, meaning the creator can turn it off if they really want lockup to mean no movement.

Reply:
I agree, one of the big reasons we need transfers though is vested token distributions. For instance, a creator wants to distribute locked tokens that their supporters can unlock in a year.